qemu/tests/pci_bridge_hotplug_unplug.py

```diff
@@ -0,0 +1,202 @@
+    images.pop(0)
+    for image in images:
+        params["dd_if"] = "ZERO"
+        params["dd_of"] = image
+        utils_test.run_virt_sub_test(test, params, env, sub_test_type)
+
+        params["dd_if"] = image
+        params["dd_of"] = "NULL"
+        utils_test.run_virt_sub_test(test, params, env, sub_test_type)
+    logging.info("Read and write data on all hotplugged disks succeed!")
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    [pci-bridge] Hotplug and unplug different devices to 1 pci-bridge
+    [pci-bridge] Hotplug and unplug different devices to different pci-bridges
+    This case will:
+    1) Start the guest.
+    2) Hotplug different devices to pci-bridge(s).
+    3) Check devices.
+    4) Hot-unplug devices.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    error_context.context("Get the main VM", logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    login_timeout = int(params.get("login_timeout", 360))
+    session = vm.wait_for_login(timeout=login_timeout)
+
+    image_list = params.get("images").split()
+    image_num = int(params.get("image_num"))
+    if len(image_list) < image_num + 1:
+        raise exceptions.TestFail("No sufficient pre-prepared images!")
+
+    device_list = []
+    one_bridge = params.get("one_bridge", True)
+    if one_bridge:
+        pci_bridge_id = params.get("pci_controllers").strip()
+
+        error_context.context("Hotplug 1 virtio-blk disk!", logging.info)
+        image = image_list.pop()
+        image_params = params.object_params(image)
+        image_filename = storage.get_image_filename(image_params, data_dir.get_data_dir())
+        image_extra_params = params.get("blk_extra_params_%s" % image)
+        virtio_blk_device = hotplug_virtio_blk_disk(vm, image_filename, pci_bridge_id, image_extra_params)
+        device_list.append(virtio_blk_device)
+        logging.info("Hotplug virtio-blk disk succeed!")
+
+        error_context.context("Hotplug 1 virtio-scsi disk!", logging.info)
+        image = image_list.pop()
+        image_params = params.object_params(image)
+        image_filename = storage.get_image_filename(image_params, data_dir.get_data_dir())
+        image_extra_params = params.get("blk_extra_params_%s" % image)
+        scsi_pci_device, scsi_hd_device = hotplug_scsi_disk(vm, image_filename, pci_bridge_id, image_extra_params)
+        device_list.append(scsi_pci_device)
+        device_list.append(scsi_hd_device)
+        logging.info("Hotplug scsi-hd disk succeed!")
+
+        error_context.context("Hotplug 1 usb disk!", logging.info)
+        image = image_list.pop()
+        image_params = params.object_params(image)
+        image_filename = storage.get_image_filename(image_params, data_dir.get_data_dir())
+        image_extra_params = params.get("blk_extra_params_%s" % image)
```

Comment:
Define a new function named get_image_filename_and_extra_params to handle this duplicated code part, can you help to re-check?